File: libreoffice.go
package excel2pdf

import (
	"fmt"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// convertExcelToPDFWithLibreOffice converts excelFilePath to PDF by invoking
// LibreOffice in headless mode. The PDF is written to the same directory as
// the source file and only the first page is retained. The path of the
// generated PDF is returned.
func convertExcelToPDFWithLibreOffice(excelFilePath string) (pdfFilePath string, err error) {
	libreOfficePath, err := findLibreOfficeBinPath()
	if err != nil {
		return "", err
	}
	// make excelFilePath absolute path
	excelFilePath, err = filepath.Abs(excelFilePath)
	if err != nil {
		slog.Error("get absolute path", "error", err, "excel_file_path", excelFilePath)
		return "", fmt.Errorf("failed to get absolute path: %w", err)
	}

	// Each concurrent LibreOffice process needs its own user profile directory.
	// Without this, multiple headless instances will collide on the same lock
	// file and fail to start.
	profileDir, err := os.MkdirTemp("", "lo_profile_*")
	if err != nil {
		return "", fmt.Errorf("failed to create libreoffice profile dir: %w", err)
	}
	defer func() {
		if err := os.RemoveAll(profileDir); err != nil {
			slog.Warn("failed to remove libreoffice profile dir", "error", err, "dir", profileDir)
		}
	}()

	// Build a file:// URL from the profile path (works on Windows and Linux).
	slashed := filepath.ToSlash(profileDir)
	if !strings.HasPrefix(slashed, "/") {
		slashed = "/" + slashed // Windows: C:/... → /C:/...
	}
	profileURL := "file://" + slashed

	cmd := exec.Command( //nolint:gosec // G204: libreOfficePath is resolved internally via findLibreOfficeBinPath, not user input
		libreOfficePath,
		"--headless",
		"--norestore",
		"--env:UserInstallation="+profileURL,
		"--convert-to", "pdf",
		"--outdir", filepath.Dir(excelFilePath),
		excelFilePath,
	) //#nosec G204

	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		slog.Error("libreoffice running", "error", err, "libre_office_path", libreOfficePath)
		return "", fmt.Errorf("failed to convert file: %w", err)
	}
	if cmd.Err != nil {
		slog.Error("libreoffice command", "error", err, "libre_office_path", libreOfficePath)
	}

	const pdfSuffix = ".pdf"
	pdfFilePath = filepath.Join(
		filepath.Dir(excelFilePath),
		fmt.Sprintf("%s%s",
			strings.TrimSuffix(
				filepath.Base(excelFilePath),
				filepath.Ext(excelFilePath),
			),
			pdfSuffix,
		),
	)

	// open the generated PDF file and delete all but the first page. Then save
	// the modified PDF file with the same name, overwriting the original PDF file.
	if err := removeAllButFirstPage(pdfFilePath); err != nil {
		slog.Error("remove all but first page", "error", err, "pdf_file_path", pdfFilePath)
		return "", fmt.Errorf("failed to remove all but first page: %w", err)
	}

	return pdfFilePath, nil
}

File: src/libreoffice.rs
use std::path::{Path, PathBuf};
use std::process::Command;

use crate::error::Excel2PdfError;

/// Converts an Excel file to PDF using LibreOffice in headless mode.
///
/// The generated PDF is written to the same directory as the source file.
/// Only the first page is retained in the output PDF.
pub fn convert_with_libreoffice(excel_path: &Path) -> crate::Result<PathBuf> {
    let lo_path = find_libreoffice_bin()?;

    let excel_path = excel_path.canonicalize().map_err(|e| {
        Excel2PdfError::ConversionFailed(format!("failed to get absolute path: {}", e))
    })?;

    let out_dir = excel_path
        .parent()
        .ok_or_else(|| Excel2PdfError::ConversionFailed("no parent directory".into()))?;

    // Each concurrent LibreOffice process needs its own user-profile directory;
    // without this, multiple headless instances collide on the same lock file.
    let profile_dir = tempfile::tempdir().map_err(|e| {
        Excel2PdfError::ConversionFailed(format!("failed to create temp dir: {}", e))
    })?;

    let profile_url = dir_to_file_url(profile_dir.path());

    let status = Command::new(&lo_path)
        .args([
            "--headless",
            "--norestore",
            &format!("--env:UserInstallation={}", profile_url),
            "--convert-to",
            "pdf",
            "--outdir",
        ])
        .arg(out_dir)
        .arg(&excel_path)
        .status()
        .map_err(|e| {
            Excel2PdfError::ConversionFailed(format!("failed to run LibreOffice: {}", e))
        })?;

    if !status.success() {
        return Err(Excel2PdfError::ConversionFailed(format!(
            "LibreOffice exited with status {}",
            status
        )));
    }

    // Construct the expected PDF path (same stem, .pdf extension).
    let stem = excel_path
        .file_stem()
        .ok_or_else(|| Excel2PdfError::ConversionFailed("file has no stem".into()))?;

    let pdf_path = out_dir.join(stem).with_extension("pdf");

    // Trim to first page only.
    crate::pdf::remove_all_but_first_page(&pdf_path)?;

    Ok(pdf_path)
}

/// Convert a directory path to a `file://` URL suitable for LibreOffice's
/// `--env:UserInstallation` argument.
fn dir_to_file_url(path: &Path) -> String {
    let mut s = path.to_string_lossy().replace('\\', "/");
    if !s.starts_with('/') {
        // Windows: `C:/...` → `/C:/...`
        s = format!("/{}", s);
    }
    format!("file://{}", s)
}

// ---------------------------------------------------------------------------
// Platform-specific resolution of the LibreOffice binary path
// ---------------------------------------------------------------------------

#[cfg(any(target_os = "linux", target_os = "macos"))]
pub fn find_libreoffice_bin() -> crate::Result<PathBuf> {
    use std::env;

    if let Ok(val) = env::var("LIBREOFFICE_PATH") {
        return Ok(PathBuf::from(val));
    }

    // Try `which libreoffice`, then `which libreoffice24.8`.
    for name in &["libreoffice", "libreoffice24.8"] {
        if let Some(p) = which_cmd(name) {
            return Ok(p);
        }
    }

    Err(Excel2PdfError::LibreOfficeNotInstalled)
}

#[cfg(any(target_os = "linux", target_os = "macos"))]
fn which_cmd(name: &str) -> Option<PathBuf> {
    let out = Command::new("which").arg(name).output().ok()?;
    if out.status.success() {
        let path = String::from_utf8_lossy(&out.stdout).trim().to_string();
        if !path.is_empty() {
            return Some(PathBuf::from(path));
        }
    }
    None
}

#[cfg(target_os = "windows")]
pub fn find_libreoffice_bin() -> crate::Result<PathBuf> {
    use std::env;
    use crate::windows::find_libreoffice_in_registry;

    if let Ok(val) = env::var("LIBREOFFICE_PATH") {
        return Ok(PathBuf::from(val));
    }

    find_libreoffice_in_registry().map_err(|e| {
        Excel2PdfError::ConversionFailed(format!("registry error: {}", e))
    })?.ok_or(Excel2PdfError::LibreOfficeNotInstalled)
}
